contracts/src/GaspToken.sol

```diff
@@ -1,30 +1,168 @@
+        if (!isAdded) {
+            revert SenderAlreadyWhitelisted(sender);
+        }
+
+        emit SenderWhitelisted(sender);
+    }
+
+    function whitelistRecipient(address recipient) external override onlyOwner {
+        if (recipient == address(0)) {
+            revert ZeroRecipient();
+        }
+
+        bool isAdded = _recipientWhitelist.add(recipient);
+        if (!isAdded) {
+            revert RecipientAlreadyWhitelisted(recipient);
+        }
+
+        emit RecipientWhitelisted(recipient);
+    }
+
+    function dewhitelistSender(address sender) external override onlyOwner {
+        if (sender == address(0)) {
+            revert ZeroSender();
+        }
+
+        bool isRemoved = _senderWhitelist.remove(sender);
+        if (!isRemoved) {
+            revert SenderNotWhitelisted(sender);
+        }
 
-contract ArbToken is MintableERC20 {
-    constructor() MintableERC20("CrossChain Token", "CCXT") {}
-}
\ No newline at end of file
+        emit SenderDewhitelisted(sender);
+    }
+
+    function dewhitelistRecipient(address recipient) external override onlyOwner {
+        if (recipient == address(0)) {
+            revert ZeroRecipient();
+        }
+
+        bool isRemoved = _recipientWhitelist.remove(recipient);
+        if (!isRemoved) {
+            revert RecipientNotWhitelisted(recipient);
+        }
+
+        emit RecipientDewhitelisted(recipient);
+    }
+
+    function getSenderWhitelist() external view override returns (address[] memory) {
+        return _senderWhitelist.values();
+    }
+
+    function getRecipientWhitelist() external view override returns (address[] memory) {
+        return _recipientWhitelist.values();
+    }
+
+    function approve(address spender, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isApproveOperationForbidden(_msgSender(), spender)
+        returns (bool)
+    {
+        return super.approve(spender, amount);
+    }
+
+    function transfer(address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isTransferOperationForbidden(_msgSender(), recipient)
+        returns (bool)
+    {
+        return super.transfer(recipient, amount);
+    }
+
+    function transferFrom(address owner, address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isTransferFromOperationForbidden(owner, _msgSender(), recipient)
+        returns (bool)
+    {
+        return super.transferFrom(owner, recipient, amount);
+    }
+
+    function allowance(address owner, address spender) public view override(ERC20, IERC20) returns (uint256) {
+        return !allowTransfers && !_checkOwnerOrSpenderWhitelisted(owner, spender) ? 0 : super.allowance(owner, spender);
+    }
+
+    function _checkOwnerOrSpenderWhitelisted(address owner, address spender) private view returns (bool) {
+        return (_senderWhitelist.contains(owner) || _senderWhitelist.contains(spender));
```

Comment:
Technically, an ERC token can be transferred in two ways:
1. A direct transfer between two parties: the sender and the recipient.
2. An indirect transfer between three parties: the original owner, the spender delegated to transfer token by the owner and the recipient. However, we still have here the same direct model where the spender is the _active sender_ on behalf of the owner who can be treated as the _passive sender_.

As we discussed with business side, the spender will be just a few entities like Uniswap V3 protocol, the Rolldown contract and the L1 council. Yes, the introduction of a separate spender list will create more control of course, but at the same will enforce more management hassle (like new methods to handle addition/removal operations and separate condition control for transfer/transferFrom operations). I think it will make sense when we intend to have many whitelisted spenders (aka active senders for indirect transfers) and what to distinguish them from senders (aka passive senders for indirect transfers). But as far as I see during our discussions with business, these lists will be very small.